<br>

### Phage annotations

<br>

**Tab.1:** Gene annotation of contigs based on Hmmer and Prodigal, using  [this database](ftp://ftp.ebi.ac.uk/pub/databases/metagenomics/viral-pipeline/hmmer_databases/pvogs.tar.gz)


```{r, echo=FALSE}
suppressMessages(library(DT))
suppressMessages(library(dplyr))
suppressMessages(library(tidyr))

annotationtable <- "#RESULTSENV#"
sample_name <- "#NAMEENV#"

dv = read.table(file = annotationtable, sep = '\t', header = FALSE)
colnames(dv) <- c("gene_prediction","contig_name","start","end","category")
datatable(dv, rownames = FALSE, class = 'hover', extensions  = 'Buttons', options = list(
    dom = 'Bfrtip', buttons = c('copy', 'csv', 'excel', 'pdf', 'colvis')))
```

#### **Visual representation of annotated genes via chromoMap**

<style>
div.blue { background-color:#e6f0ff; border-radius: 5px; padding: 20px;}
</style>
<div class = "blue">
Visual annotation of phage contigs and annotated protein-coding genes via chromoMap is stored by default here:   
```shell
results/your_sample/sample_overview_large.html
results/your_sample/sample_overview_small.html
```

</div>


<a href="#top">Back to top</a>